### PR TITLE
Add final-form and react-final-form packages, and create Config Table sample

### PR DESCRIPTION
**Pull request by AndreaLandiArk**
This pull request introduces enhancements to the configuration table using Final Form with a Chakra UI style. Notable features include the ability to add and remove items, single-row validation, and across-rows validation.

Note:
The submission process currently simulates field serialization into a JSON object for testing purposes. In the final implementation, the entire configuration will be sent collectively to the right endpoint for the TOTAL replace (our config behaviour).

image

### CODE REVIEW

File: src/features/configTable/configTable.tsx [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  configData: Employee[];
+};
+
+export type Employee = {
+  name: string;
+  surName: string;
+  employed: boolean;
+};
+
+const onSubmit = async (values: { configData: Employee[] }) => {
```
Comment (AndreaCuneo) [inline src/features/configTable/configTable.tsx:38]: please show how to wire it api mutations since the use case is to call into a remote api. 

you can use a custom queryFn to noop in RTKQ.

Add a delay of 2sec to simulate network and show how to handle disable buttons, spinners and all the needed details to make it a real case.

https://redux-toolkit.js.org/rtk-query/usage/customizing-queries#using-a-no-op-queryfn

File: src/features/configTable/configTable.tsx [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+                          <Td>
+                            <Field
+                              name={`${employee}.name`}
+                              render={({ input, meta }) => (
+                                <Box>
+                                  <FormControl>
+                                    <Input
+                                      {...input}
+                                      placeholder="First Name"
+                                      isInvalid={
```
Comment (AndreaCuneo) [inline src/features/configTable/configTable.tsx:145]: this 'is invalid' is repeaded too many times. also if I remember the library, it should be taken from meta.

Please add Chakra control helpers as I did which factor this out.

File: src/features/configTable/configTable.tsx [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                  <Th>Employed</Th>
+                  <Th>Actions</Th>
+                </Tr>
+              </Thead>
+              <Tbody>
+                <FieldArray name="configData">
+                  {({ fields }) =>
+                    fields
+                      .map((employee, index) => (
+                        <Tr key={employee}>
```
Comment (AndreaCuneo) [inline src/features/configTable/configTable.tsx:135]: these 3 lines are quite strange.

you have a 'fields' list, each entry is an 'employee' (so employee is a field?) and employee which looks like an 'object' is used as a Key and below as a string. 

either this doesn't work, or the naming of variables is quite hard to follow.

File: src/features/configTable/configTable.tsx [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+  name: string;
+  surName: string;
+  employed: boolean;
+};
+
+const onSubmit = async (values: { configData: Employee[] }) => {
+  window.alert(JSON.stringify(values));
+};
+
+const validateForm = (values: { configData: Employee[] }) => {
```
Comment (AndreaCuneo) [inline src/features/configTable/configTable.tsx:42]: did you evaluated a Validation library like Zod or Yup?

**Review by AndreaCuneo — CHANGES_REQUESTED** [review]

File: src/features/configTable/configTable.tsx (hunk 2 above)
Comment (AndreaLandiArk) [reply src/features/configTable/configTable.tsx]: isInvalid is the standard for Chackra.
The control helper <FormErrorMessage> is not working because it should be used in combination with "meta" that we do not use since we have per row validation.
<FormHelperText> is still a Chackra control helper originally not designed to be an error but it differs only for the color.

File: src/features/configTable/configTable.tsx (hunk 4 above)
Comment (AndreaLandiArk) [reply src/features/configTable/configTable.tsx:44]: I think there is a task for this specifically but is not in the scope of this one.

File: src/features/configTable/configTableExample.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { Box } from "@chakra-ui/react";
+import ConfigTable, { Employee } from "./configTable";
+import { useGetConfigQuery } from "./configTableApi";
+
+const ConfigTableExampleView = () => {
+  const { data, isLoading: getConfigIsLoading } = useGetConfigQuery(null, {
```
Comment (AndreaCuneo) [inline src/features/configTable/configTableExample.tsx:6]: I do expect get and post to be at the same 'level'. in this case, I expect the get to be done inside the EmployeeConfigTable

File: src/lib/zod2form.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import * as z from "zod";
+import _ from "lodash";
+
+/*
```
Comment (AndreaCuneo) [inline src/lib/zod2form.ts:4]: this comment is not in line with the code below. so either the code or the comment are wrong.

the code generate:

{ "table": [ { "name": ["error 1"] } ] }

File: src/features/configTable/configTableApi.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+export const configTableApiSlice = createApi({
+  reducerPath: "configTableApi",
+  baseQuery: fetchBaseQuery({
+    baseUrl: "https://myconfigUrlApi.com/",
+  }),
+  tagTypes: ["Employee"],
+  endpoints: (builder) => ({
+    getConfig: builder.query<Employee[], null>({
+      queryFn: async () => {
+        return { data };
```
Comment (AndreaCuneo) [inline src/features/configTable/configTableApi.ts:36]: add a sleep to simulate data load and load spinners

File: src/features/configTable/configTableApi.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      },
+      providesTags: ["Employee"],
+    }),
+    postConfig: builder.mutation<void, Employee[]>({
+      queryFn: async (body) => {
+        return new Promise((resolve) => {
```
Comment (AndreaCuneo) [inline src/features/configTable/configTableApi.ts:42]: this is a 'nice' lib, move it to lib/

function delay(ms: number) {
    return new Promise( resolve => setTimeout(resolve, ms) );
}

and change the code to

await delay(2000);
return { data: undefined }

File: src/features/configTable/configTable.tsx [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+  };
+
+export default function EditableTableExample(props: {
+  title: string;
+  data: Employee[] | undefined;
+  isLoading: boolean;
+}) {
+  const [
+    postConfig,
+    { isLoading: postConfigIsLoading, isSuccess: postConfigSuccess },
```
Comment (AndreaCuneo) [inline src/features/configTable/configTable.tsx:115]: test a Failure case, in case the Submit fails. you can use 2 separate buttons to call postConfig with 2 different args


await postConfig(values.table, true).unwrap()
or
await postConfig(values.table, false).unwrap()

**Review by AndreaCuneo — CHANGES_REQUESTED** [review]

File: src/features/configTable/configTable.tsx [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+                type="submit"
+                isDisabled={submitting || pristine || hasValidationErrors}
+                isLoading={submitting || postConfigIsLoading}
+              >
+                Submit
+              </Button>
+              <Button
+                isDisabled={submitting || pristine || hasValidationErrors}
+                isLoading={submitting || postConfigIsLoading}
+                onClick={() => onSubmitError(form.getState().values)}
```
Comment (AndreaCuneo) [inline src/features/configTable/configTable.tsx:201]: this is not ok as the Form doesn't know that you're submitting. you need to call into handleSubmit.

check FF documentation but this is not ok.

as an alternative, you can add a top level element (alongside the table), which is checked by the postHandler to fail.
like a checkbox.

**Review by AndreaCuneo — CHANGES_REQUESTED** [review]

File: src/features/configTable/configTable.tsx [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+      }}
+    />
+  );
+}
+
+const TableRow = (props: {
+  name: string;
+  index: number;
+  submitting: boolean;
+  fields: any;
```
Comment (AndreaCuneo) [inline src/features/configTable/configTable.tsx:262]: why fields?

File: src/features/configTable/configTable.tsx [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+                onClick={() => {
+                  form.change("throwError", false);
+                }}
+              >
+                Submit
+              </Button>
+              <Button
+                type="submit"
+                isDisabled={submitting || pristine || hasValidationErrors}
+                isLoading={submitting || postConfigIsLoading}
```
Comment (AndreaCuneo) [inline src/features/configTable/configTable.tsx:204]: only submitting should be needed.

**Review by AndreaCuneo — APPROVED** [review]
2 minor cleanups, overall is ok.

File: src/features/configTable/configTable.tsx (hunk 11 above)
Comment (AndreaLandiArk) [reply src/features/configTable/configTable.tsx:262]: for the delete

Comment (AndreaLandiArk) [reply src/features/configTable/configTable.tsx:262]:
```typescript
 <IconButton
          icon={<FaTrash />}
          onClick={() => fields.remove(index)}
          colorScheme="red"
          aria-label="Remove Employee"
        />
      </Td>
```

File: src/features/configTable/configTable.tsx [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+            <FormControl isInvalid={error && touched} isDisabled={submitting}>
+              <Input {...input} />
+              <FormErrorMessage>{error}</FormErrorMessage>
+            </FormControl>
+          )}
+        />
+      </Td>
+      <Td>
+        <Field
+          name={`${name}.employed`}
```
Comment (sean-sligo) [inline src/features/configTable/configTable.tsx:300]: add type="checkbox"

**Review by sean-sligo — APPROVED** [review]

**Review by sean-sligo — APPROVED** [review]
